File: code_snippets/go/usecases/microservices/saga/subscriptionsaga.go
package saga

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type SubscriptionRequest struct {
	UserID        string   `json:"userId"`
	CreditCard    string   `json:"creditCard"`
	Subscriptions []string `json:"subscriptions"`
}

// <start_here>
type SubscriptionSaga struct{}

func (SubscriptionSaga) Add(ctx restate.Context, req SubscriptionRequest) (err error) {
	// <mark_1>
	compensations := make([]func() error, 0, len(req.Subscriptions)+1)
	// </mark_1>
	// Run compensations at the end if err != nil
	defer func() {
		if err != nil {
			// <mark_2>
			for _, compensation := range compensations {
				if _, compErr := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
					return restate.Void{}, compensation()
				}); compErr != nil {
					err = compErr
				}
			}
			// </mark_2>
		}
	}()

	paymentId := restate.Rand(ctx).UUID().String()

	// <mark_1>
	compensations = append(compensations, func() error {
		return RemoveRecurringPayment(req.CreditCard, paymentId)
	})
	// </mark_1>
	// <mark_1> green
	_, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return CreateRecurringPayment(req.CreditCard, paymentId)
		// </mark_1>
	})
	if err != nil {
		return
	}

	for _, subscription := range req.Subscriptions {
		// <mark_1>
		compensations = append(compensations, func() error {
			return RemoveSubscription(req.UserID, subscription)
		})
		// </mark_1>
		// <mark_1> green
		if _, err = restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
			return CreateSubscription(req.UserID, subscription)
			// </mark_1>
		}); err != nil {
			return
		}
	}
	return nil
}

// <end_here>

func main() {
	if err := server.NewRestate().
		Bind(restate.Reflect(SubscriptionSaga{})).
		Start(context.Background(), ":9080"); err != nil {

		log.Fatal(err)
	}
}

func RemoveSubscription(id string, subscription string) error {
	return nil
}

func RemoveRecurringPayment(card string, id string) error {
	return nil
}

func CreateSubscription(id string, subscription string) (restate.Void, error) {
	return restate.Void{}, nil
}

func CreateRecurringPayment(card string, id string) (restate.Void, error) {
	return restate.Void{}, nil
}

File: code_snippets/kotlin/src/main/kotlin/usecases/microservices/saga/SubscriptionSaga.kt
package usecases.microservices.saga

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.Context
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlinx.serialization.Serializable

// <start_here>
@Service
class SubscriptionSaga {

  @Handler
  suspend fun add(ctx: Context, req: SubscriptionRequest) {
    // !focus(1:17)
    // <mark_1>
    val compensations: MutableList<suspend () -> Unit> = mutableListOf()
    // </mark_1>

    try {
      val paymentId = ctx.random().nextUUID().toString()
      // <mark_1>
      compensations.add { removeRecurringPayment(paymentId) }
      // </mark_1>
      // <mark_1> green
      ctx.runBlock { createRecurringPayment(req.creditCard, paymentId) }
      // </mark_1>

      for (subscription in req.subscriptions) {
        // <mark_1>
        compensations.add { removeSubscription(req.userId, subscription) }
        // </mark_1>
        // <mark_1> green
        ctx.runBlock { createSubscription(req.userId, subscription) }
        // </mark_1>
      }
    } catch (e: TerminalException) {
      // <mark_2>
      compensations.reversed().forEach { ctx.runBlock { it() } }
      // </mark_2>
      throw TerminalException("Subscription failed")
    }
  }
}
// <end_here>

fun main() {
  RestateHttpServer.listen(endpoint { bind(SubscriptionSaga()) })
}

@Serializable
data class SubscriptionRequest(
    val userId: String,
    val creditCard: String,
    val subscriptions: List<String>
)

fun createSubscription(userId: String, subscription: String): String {
  return "SUCCESS"
}

fun createRecurringPayment(creditCard: String, paymentId: String): String {
  return "payment-reference"
}

fun removeSubscription(userId: String, subscription: String) {
  return
}

fun removeRecurringPayment(paymentId: String) {
  return
}
